quinn-proto/src/congestion/bbr.rs

```diff
@@ -0,0 +1,653 @@
+    }
+    fn on_ack(
+        &mut self,
+        now: Instant,
+        sent: Instant,
+        bytes: u64,
+        app_limited: bool,
+        rtt: &RttEstimator,
+    ) {
+        self.bbr_state.max_bandwidth.on_ack(
+            now,
+            sent,
+            bytes,
+            self.bbr_state.round_count,
+            app_limited,
+        );
+        self.bbr_state.acked_bytes += bytes;
+        if self.bbr_state.is_min_rtt_expired(now, app_limited) || self.bbr_state.min_rtt > rtt.min()
+        {
+            self.bbr_state.min_rtt = rtt.min();
+        }
+    }
+
+    fn on_end_acks(
+        &mut self,
+        now: Instant,
+        in_flight: u64,
+        app_limited: bool,
+        largest_packet_num_acked: Option<u64>,
+    ) {
+        let bytes_acked = self.bbr_state.max_bandwidth.bytes_acked_this_window();
+        let excess_acked = self.bbr_state.ack_aggregation.update_ack_aggregation_bytes(
+            bytes_acked,
+            now,
+            self.bbr_state.round_count,
+            self.bbr_state.max_bandwidth.get_estimate(),
+        );
+        self.bbr_state
+            .max_bandwidth
+            .end_acks(self.bbr_state.round_count, app_limited);
+        if let Some(largest_acked_packet) = largest_packet_num_acked {
+            self.bbr_state.max_acked_packet_number = largest_acked_packet;
+        }
+
+        let mut is_round_start = false;
+        if bytes_acked > 0 {
+            is_round_start = self.bbr_state.max_acked_packet_number
+                > self.bbr_state.current_round_trip_end_packet_number;
+            if is_round_start {
+                self.bbr_state.current_round_trip_end_packet_number =
+                    self.bbr_state.max_sent_packet_number;
+                self.bbr_state.round_count += 1;
+            }
+        }
+
+        self.bbr_state.update_recovery_state(is_round_start);
+
+        if self.bbr_state.mode == Mode::ProbeBw {
+            self.bbr_state.update_gain_cycle_phase(now, in_flight);
+        }
+
+        if is_round_start && !self.bbr_state.is_at_full_bandwidth {
+            self.bbr_state.check_if_full_bw_reached(app_limited);
+        }
+
+        self.bbr_state.maybe_exit_startup_or_drain(now, in_flight);
+
+        self.bbr_state
+            .maybe_enter_or_exit_probe_rtt(now, is_round_start, in_flight, app_limited);
+
+        // After the model is updated, recalculate the pacing rate and congestion window.
+        self.bbr_state.calculate_pacing_rate();
+        self.bbr_state.calculate_cwnd(bytes_acked, excess_acked);
+        self.bbr_state.calculate_recovery_window(
+            bytes_acked,
+            self.bbr_state.loss_state.lost_bytes,
+            in_flight,
+        );
+
+        self.bbr_state.prev_in_flight_count = in_flight;
+        self.bbr_state.loss_state.reset();
+    }
+
+    fn on_loss(&mut self, _now: Instant, _sent: Instant, bytes: u64) {
+        self.bbr_state.loss_state.lost_bytes += bytes;
+    }
```

Comment:
I'm surprised BBR completely ignores ECN. You're certain that's intended?

Reply:
yes. BBR ignores it. BBR2 does not.